content/en/blog/_posts/2024-02-29-cri-o-seccomp-oci-artifacts.md

```diff
@@ -0,0 +1,486 @@
+INFO[…] Pulling OCI artifact from ref: quay.io/crio/seccomp:v1  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Trying to read profile from: /var/lib/containers/storage/overlay/c15f3eb1efd9abe0ae973e050c374988bba52c45a53f5c3cd80419a865013fa6/merged/seccomp.json  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Retrieved OCI artifact seccomp profile of len: 18853  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+INFO[…] Created container 116a316cd9a11fe861dd04c43b94f45046d1ff37e2ed05a4e4194fcaab29ee63: default/pod/container  id=c1f22c59-e30e-4046-931d-a0c0fdc2c8b7 name=/runtime.v1.RuntimeService/CreateContainer
+```
+
+```shell
+export CONTAINER_ID=$(sudo crictl ps --name container -q)
+sudo crictl inspect $CONTAINER_ID | jq .info.runtimeSpec.linux.seccomp | head
+```
+
+```yaml
+{
+  "defaultAction": "SCMP_ACT_ERRNO",
+  "defaultErrnoRet": 38,
+  "architectures": [
+    "SCMP_ARCH_X86_64",
+    "SCMP_ARCH_X86",
+    "SCMP_ARCH_X32"
+  ],
+  "syscalls": [
+    {
+```
+
+For container images, the annotation `seccomp-profile.kubernetes.cri-o.io` will
+be treated in the same way as `seccomp-profile.kubernetes.cri-o.io/POD` and
+applies to the whole pod. In addition to that, the whole feature also works when
+using the container specific annotation on an image, for example if a container
+is named `container1`:
+
+```shell
+skopeo inspect --raw docker://quay.io/crio/nginx-seccomp:container |
+    jq '.annotations."seccomp-profile.kubernetes.cri-o.io/container1"'
+```
+
+```console
+"quay.io/crio/seccomp:v1"
+```
+
+The cool thing about this whole feature is that users can now create seccomp
+profiles for specific container images and store them side by side in the same
+registry. Linking the images to the profiles provides a great flexibility to
+maintain both over the whole application's life cycle.
+
+### Pushing profiles using ORAS
+
+The actual creation of the profiles requires a bit more prework, but I have the
+hope that future tools will simplify the overall process. Right now, the
+container registry needs to be [ORAS
+compatible](https://oras.land/docs/compatible_oci_registries/#registries-supporting-oci-artifacts),
+which is also (partially) the case for [Quay.io](https://quay.io). CRI-O also
+needs a `config` with the `mediaType` `application/vnd.oci.image.config.v1+json`,
+while ORAS uses `application/vnd.oci.empty.v1+json` per default. Last but not
+least, CRI-O requires a tarball rather than the plain `.json` file. To achieve
+all of that, the following commands can be executed:
+
+```shell
+echo "{}" > config.json
+tar cf seccomp.tar seccomp.json
+oras push \
+    --config config.json:application/vnd.oci.image.config.v1+json \
```

Comment:
Is this use of `application/vnd.oci.image.config.v1+json`:
- the final design?
- a prototype for a future approach that will be different from the prototype?

I'd like to make that clear to readers, too.

For example, Helm uses `application/vnd.cncf.helm.config.v1+json` with `application/vnd.cncf.helm.chart.content.v1.tar+gzip` layers.

Reply:
For ORAS it would be the final design (for v1.30). I expect support for artifacts in Podman in the future which could then simplify things.

Reply:
:grimacing: 

I'm not a fan; getting a dedicated media type assigned feels like the obvious right approach (so that you can tell a seccomp profile apart from a container image).

There's also the precedent it sets: encouraging other projects to further use the container image media type for other things that you can't actually execute with `podman run`. Should we, Kubernetes, be promoting this kind of misunderstanding - even indirectly?

---

If we cast this as an _experiment_, all my qualms about the approach fall away. Experiments are fine to cut corners and readers would be unwise to treat them as good practice.

Reply:
(eg, media type is not really assigned)
```shell
    --config config.json:application/vnd.cncf.seccomp-profile.config.v1+json \
```

Reply:
Generally SGTM, but I'm not sure if we can merge / resolve that until next week in https://github.com/containers/image/pull/2306. Would you be fine in mentioning that work and outlining the future path?

Reply:
Pretty much yes. You could:
- frame this as experimental / incomplete support
- explain that the media type thing is likely to change
- extend the “future steps” part of the article to explain the likely journey specifically around media types and ORAS

or pick multiple items from the above.

Reply:
:cool:

...but, the object still has _some_ media type, right? Happy to have a call about the approach here, both in terms of the CRI-O level and how to follow the OCI good practice, or about how to caveat any remaining gaps.

Reply:
Ah yes we still enforce the media type `application/vnd.cncf.seccomp-profile.config.v1+json`.

Reply:
See https://github.com/kubernetes/website/pull/45121#issuecomment-1966071714